pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/json/ImmutableJsonIndexReader.java

```diff
@@ -300,6 +305,92 @@ private int getDocId(int flattenedDocId) {
     return _docIdMapping.getInt((long) flattenedDocId << 2);
   }
 
+  /**
+   * ImmutableJsonIndexReaderContext holds a cache that is used to accelerate following reads. The cache is specific
+   * to a key, and therefore should NOT be reused for other keys. For each key, ImmutableJsonIndexReader.createContext()
+   * should be invoked to create a fresh context.
+   */
+  public static class ImmutableJsonIndexReaderContext implements JsonIndexReaderContext {
+    private final Int2ObjectOpenHashMap<RoaringBitmap> _cache;
+
+    public ImmutableJsonIndexReaderContext() {
+      _cache = new Int2ObjectOpenHashMap<>();
+    }
+
+    public Int2ObjectOpenHashMap<RoaringBitmap> getCache() {
+      return _cache;
+    }
+  }
+
+  @Override
+  public ImmutableJsonIndexReaderContext createContext() {
+    return new ImmutableJsonIndexReaderContext();
+  }
+
+  @Override
+  public String[] getValuesForKeyAndDocs(String key, int[] docIds, JsonIndexReaderContext context) {
```

Comment:
I wouldn't model it as a context class since it is not sharable across keys. Suggest adding a separate method `public Int2ObjectOpenHashMap<RoaringBitmap> getMatchingDocsMap(String key)`, then pass this map into this method

Reply:
I went w/ a context class since the 'cache' is a different type for the mutable/immutable segment implementation. 

I think having a version of the method for immutable/mutable impls, `public List<RoaringBitmap> getMatchingDocsMap(String key)` and `public Map<String, RoaringBitmap> getMatchingDocsMap(String key)` would move the handling of this behavior outside the IndexReader. This would have downsides of requiring that the caller of JsonIndexReader have more insight into the implementation of JsonIndexReader, and expose the cache object to the caller. Let me know if I'm missing something.

Reply:
The main concern is that the context here is associated with a key instead of the index. This can cause confusion because it is not used the same way as the forward index context, and creating context doesn't even take the key as the argument.
Taking another look, seems we can always use `Map<String, RoaringBitmap>` as the context and avoid extra dictionary lookup. Basically using `public Map<String, RoaringBitmap> getMatchingDocsMap(String key)` for both immutable and mutable index

Reply:
Gotcha, thanks for the detailed explanation. Refactored to use `public Map<String, RoaringBitmap> getMatchingDocsMap(String key)` as you suggested.